# Foundations of Computer Science Project 2024/2024
Maliwat Julius [864520]


### Project Setup

Before proceeding with the execution of the notebook:  
- The first cell installs all the dependencies listed in `requirements.txt`.  
- The second cell runs a script to download and extract the dataset into the `data` folder.

These steps need to be performed only once.

**Note**: This project is developed using Python version `3.12.1`.


In [1]:
%pip install -r requirements.txt --quiet

Note: you may need to restart the kernel to use updated packages.


In [2]:
!python scripts/download_data.py

Download successful.
Files in archive: ['data.csv']
Extracting data.csv...
data.csv extracted to data.
Data is ready at data\data.csv


### Project Exec

I import the `pandas` library and define constants for handling missing values (`FILL_STRING`) and specifying the dataset path (`DATA_PATH`).

In [3]:
import pandas as pd
FILL_STRING = "N/A"
DATA_PATH = 'data/data.csv'

I load the dataset from the specified `DATA_PATH` and display its first few rows. 

In [4]:
df = pd.read_csv(DATA_PATH)
df

C:\Users\Admin\AppData\Local\Temp\ipykernel_12696\119131063.py:1: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(DATA_PATH)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0             1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   
1             1.0  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0   
2             1.0  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0   
3             1.0  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0   
4             2.0  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0   
...           ...                  ...                   ...              ...   
6405003       NaN  2020-01-31 22:51:00   2020-01-31 23:22:00              NaN   
6405004       NaN  2020-01-31 22:10:00   2020-01-31 23:26:00              NaN   
6405005       NaN  2020-01-31 22:50:07   2020-01-31 23:17:57              NaN   
6405006       NaN  2020-01-31 22:25:53   2020-01-31 22:48:32              NaN   
6405007       NaN  2020-01-31 22:44:00   2020-01-31 23:06:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 1.20         1.0                  N           238   
1                 1.20         1.0                  N           239   
2                 0.60         1.0                  N           238   
3                 0.80         1.0                  N           238   
4                 0.00         1.0                  N           193   
...                ...         ...                ...           ...   
6405003           3.24         NaN                NaN           237   
6405004          22.13         NaN                NaN           259   
6405005          10.51         NaN                NaN           137   
6405006           5.49         NaN                NaN            50   
6405007          11.60         NaN                NaN           179   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 239           1.0         6.00   3.00      0.5        1.47   
1                 238           1.0         7.00   3.00      0.5        1.50   
2                 238           1.0         6.00   3.00      0.5        1.00   
3                 151           1.0         5.50   0.50      0.5        1.36   
4                 193           2.0         3.50   0.50      0.5        0.00   
...               ...           ...          ...    ...      ...         ...   
6405003           234           NaN        17.59   2.75      0.5        0.00   
6405004            45           NaN        46.67   2.75      0.5        0.00   
6405005           169           NaN        48.85   2.75      0.0        0.00   
6405006            42           NaN        27.17   2.75      0.0        0.00   
6405007           205           NaN        54.56   2.75      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    0.3         11.27   
1                0.00                    0.3         12.30   
2                0.00                    0.3         10.80   
3                0.00                    0.3          8.16   
4                0.00                    0.3          4.80   
...               ...                    ...           ...   
6405003          0.00                    0.3         21.14   
6405004         12.24                    0.3         62.46   
6405005          0.00                    0.3         51.90   
6405006          0.00                    0.3         30.22   
6405007          0.00                    0.3         58.11   

         congestion_surcharge  
0                         2.5  
1                         2.5  
2                         2.5  
3                         0.0  
4                         0.0  
...                       ...  
6405003                   0.0  
6405004                   0.0  
6405005                   0.0  
6405006                   0.0  
6405007                   0.0  

[6405008 rows x 18 columns]

The dataset contains **6,405,008 rows** and **18 columns**. However, a `DtypeWarning` is raised, indicating that column 6 (`store_and_fwd_flag`) has mixed data types. To understand the issue, I investigate the unique values and their counts in this column using the `value_counts` method.







In [5]:
count_values = df["store_and_fwd_flag"].value_counts(dropna=False)

# Stampa i risultati
print(count_values)

store_and_fwd_flag
N      6271447
Y        68120
NaN      65441
Name: count, dtype: int64


The `store_and_fwd_flag` column contains three distinct values:
- `'N'`: 6,271,447 rows
- `'Y'`: 68,120 rows
- `NaN`: 65,441 rows

The presence of `NaN` values is likely the cause of the `DtypeWarning`, as it introduces mixed data types in the column. To address this, I fill the missing values with the constant `FILL_STRING` (`"N/A"`). After filling, I recheck the value counts to confirm the update.

In [6]:
df['store_and_fwd_flag'] = df['store_and_fwd_flag'].fillna(FILL_STRING)
count_values = df["store_and_fwd_flag"].value_counts(dropna=False)

# Stampa i risultati
print(count_values)

store_and_fwd_flag
N      6271447
Y        68120
N/A      65441
Name: count, dtype: int64


I calculate the completeness of each column. This helps identify missing data and assess its impact on the overall quality of the dataset.







In [7]:
# Calcolo della completezza per ciascuna colonna
completeness = (df.notnull().sum() / len(df)) * 100

# Trasformazione in un DataFrame per una visualizzazione più chiara
completeness_df = completeness.reset_index()
completeness_df.columns = ["Column", "Completeness (%)"]

# Stampa i risultati
print(completeness_df)


                   Column  Completeness (%)
0                VendorID         98.978284
1    tpep_pickup_datetime        100.000000
2   tpep_dropoff_datetime        100.000000
3         passenger_count         98.978284
4           trip_distance        100.000000
5              RatecodeID         98.978284
6      store_and_fwd_flag        100.000000
7            PULocationID        100.000000
8            DOLocationID        100.000000
9            payment_type         98.978284
10            fare_amount        100.000000
11                  extra        100.000000
12                mta_tax        100.000000
13             tip_amount        100.000000
14           tolls_amount        100.000000
15  improvement_surcharge        100.000000
16           total_amount        100.000000
17   congestion_surcharge        100.000000


#### 1. Extract all trips with trip_distance larger than 50



I use a boolean mask (`df["trip_distance"] > 50`) to filter rows where the `trip_distance` exceeds 50.


In [8]:
long_distance_trips = df[df["trip_distance"] > 50]
long_distance_trips.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
23842       2.0  2020-01-01 01:53:07   2020-01-01 03:54:41              1.0   
39013       2.0  2020-01-01 02:05:07   2020-01-01 03:03:10              1.0   
41620       1.0  2020-01-01 03:05:54   2020-01-01 04:16:26              1.0   
58262       2.0  2020-01-01 05:36:12   2020-01-01 06:40:06              1.0   
63024       2.0  2020-01-01 07:40:30   2020-01-01 08:40:01              1.0   

       trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
23842          52.30         5.0                  N           262   
39013          51.23         5.0                  N           264   
41620          53.80         5.0                  N           132   
58262          55.23         5.0                  N           132   
63024          54.19         5.0                  N           132   

       DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
23842           265           1.0        300.0    0.0      0.0       61.78   
39013           264           1.0        329.0    0.0      0.5      100.78   
41620           265           1.0        250.0    0.0      0.0       53.35   
58262           265           2.0        170.0    0.0      0.5        0.00   
63024           265           1.0        230.0    0.0      0.0        0.00   

       tolls_amount  improvement_surcharge  total_amount  congestion_surcharge  
23842          6.12                    0.3        370.70                   2.5  
39013          6.12                    0.3        436.70                   0.0  
41620         16.62                    0.3        320.27                   0.0  
58262         18.26                    0.3        189.06                   0.0  
63024         12.24                    0.3        242.54                   0.0

#### 2. Extract all trips where payment_type is missing



I use the `.isna()` method to create a boolean mask that identifies rows where the `payment_type` column has missing (`NaN`) values. The mask is applied to filter the dataset


In [9]:
missing_payment_trips = df[df["payment_type"].isna()]

missing_payment_trips.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
6339567       NaN  2020-01-01 08:51:00   2020-01-01 09:19:00              NaN   
6339568       NaN  2020-01-01 08:38:43   2020-01-01 08:51:08              NaN   
6339569       NaN  2020-01-01 08:27:00   2020-01-01 08:32:00              NaN   
6339570       NaN  2020-01-01 08:46:00   2020-01-01 08:57:00              NaN   
6339571       NaN  2020-01-01 08:21:00   2020-01-01 08:38:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
6339567          13.69         NaN                N/A           136   
6339568           3.42         NaN                N/A           121   
6339569           2.20         NaN                N/A           197   
6339570           0.84         NaN                N/A           262   
6339571           7.24         NaN                N/A            45   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
6339567           232           NaN        51.05   2.75      0.5         0.0   
6339568             9           NaN        27.06   2.75      0.0         0.0   
6339569           216           NaN        24.36   2.75      0.5         0.0   
6339570           236           NaN        26.08   2.75      0.5         0.0   
6339571           142           NaN        25.28   2.75      0.5         0.0   

         tolls_amount  improvement_surcharge  total_amount  \
6339567           0.0                    0.3         54.60   
6339568           0.0                    0.3         30.11   
6339569           0.0                    0.3         27.91   
6339570           0.0                    0.3         29.63   
6339571           0.0                    0.3         28.83   

         congestion_surcharge  
6339567                   0.0  
6339568                   0.0  
6339569                   0.0  
6339570                   0.0  
6339571                   0.0

#### 3. For each (PULocationID, DOLocationID) pair, determine the number of trips







I use the `.groupby()` method to group the dataset by `PULocationID` and `DOLocationID`. The `.size()` method calculates the number of trips for each unique pair, and `.reset_index(name="trip_count")` converts the result into a DataFrame with a new column, `trip_count`, that stores the counts.


In [10]:
trips_per_location_pair = df.groupby(["PULocationID", "DOLocationID"]).size().reset_index(name="trip_count")

trips_per_location_pair.head()

PULocationID  DOLocationID  trip_count
0             1             1         638
1             1            50           1
2             1            68           1
3             1           138           2
4             1           140           1

#### 4. Save all rows with missing VendorID, passenger_count, store_and_fwd_flag, payment_type in a new dataframe called bad, and remove those rows from the original dataframe.



I use `.isna().any(axis=1)` to create a boolean mask that identifies rows where at least one value is missing (`NaN`) in the specified columns: `VendorID`, `passenger_count`, `store_and_fwd_flag`, and `payment_type`. Using this mask, I extract these rows into a new DataFrame called `bad`. 

I then remove these rows from the original DataFrame using `.drop(bad.index)`.

In [11]:
bad = df[df[["VendorID", "passenger_count", "store_and_fwd_flag", "payment_type"]].isna().any(axis=1)]

df = df.drop(bad.index)
bad.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
6339567       NaN  2020-01-01 08:51:00   2020-01-01 09:19:00              NaN   
6339568       NaN  2020-01-01 08:38:43   2020-01-01 08:51:08              NaN   
6339569       NaN  2020-01-01 08:27:00   2020-01-01 08:32:00              NaN   
6339570       NaN  2020-01-01 08:46:00   2020-01-01 08:57:00              NaN   
6339571       NaN  2020-01-01 08:21:00   2020-01-01 08:38:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
6339567          13.69         NaN                N/A           136   
6339568           3.42         NaN                N/A           121   
6339569           2.20         NaN                N/A           197   
6339570           0.84         NaN                N/A           262   
6339571           7.24         NaN                N/A            45   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
6339567           232           NaN        51.05   2.75      0.5         0.0   
6339568             9           NaN        27.06   2.75      0.0         0.0   
6339569           216           NaN        24.36   2.75      0.5         0.0   
6339570           236           NaN        26.08   2.75      0.5         0.0   
6339571           142           NaN        25.28   2.75      0.5         0.0   

         tolls_amount  improvement_surcharge  total_amount  \
6339567           0.0                    0.3         54.60   
6339568           0.0                    0.3         30.11   
6339569           0.0                    0.3         27.91   
6339570           0.0                    0.3         29.63   
6339571           0.0                    0.3         28.83   

         congestion_surcharge  
6339567                   0.0  
6339568                   0.0  
6339569                   0.0  
6339570                   0.0  
6339571                   0.0

#### 5. Add a duration column storing how long each trip has taken (use tpep_pickup_datetime, tpep_dropoff_datetime)



I ensure that the `tpep_pickup_datetime` and `tpep_dropoff_datetime` columns are in datetime format using `pd.to_datetime`. This step is necessary for performing datetime operations.

I then calculate the trip duration by subtracting the pickup time (`tpep_pickup_datetime`) from the dropoff time (`tpep_dropoff_datetime`). The resulting timedelta is converted to total seconds using `.dt.total_seconds()` and divided by 60 to express the duration in minutes. 

Finally, I create a new column, `duration`, in the DataFrame to store the calculated values.

In [12]:

df["tpep_pickup_datetime"] = pd.to_datetime(df["tpep_pickup_datetime"])
df["tpep_dropoff_datetime"] = pd.to_datetime(df["tpep_dropoff_datetime"])


df["duration"] = (df["tpep_dropoff_datetime"] - df["tpep_pickup_datetime"]).dt.total_seconds() / 60

df[["tpep_pickup_datetime", "tpep_dropoff_datetime", "duration"]].head()

tpep_pickup_datetime tpep_dropoff_datetime  duration
0  2020-01-01 00:28:15   2020-01-01 00:33:03  4.800000
1  2020-01-01 00:35:39   2020-01-01 00:43:04  7.416667
2  2020-01-01 00:47:41   2020-01-01 00:53:52  6.183333
3  2020-01-01 00:55:23   2020-01-01 01:00:14  4.850000
4  2020-01-01 00:01:58   2020-01-01 00:04:16  2.300000

#### 6. For each pickup location, determine how many trips have started there.


I group the dataset by `PULocationID` using `.groupby()` and count the number of trips for each location with `.size()`. The result is converted into a DataFrame with a new column, `trip_count`, using `.reset_index(name="trip_count")`.

In [13]:
trips_per_pickup_location = df.groupby("PULocationID").size().reset_index(name="trip_count")

trips_per_pickup_location.head()


PULocationID  trip_count
0             1         753
1             2           3
2             3          70
3             4        9902
4             5          39

#### 7. Cluster the pickup time of the day into 30-minute intervals (e.g. from 02:00 to 02:30)


I create a new column, `pickup_30min_interval`, to group pickup times into 30-minute intervals.

- I calculate the rounded time once by using `.dt.floor("30min")` on `tpep_pickup_datetime` and store the result in a temporary Series called `pickup_rounded`.
- I format the rounded time (`pickup_rounded`) into `HH:MM` format using `.dt.strftime("%H:%M")`.
- To create the interval, I calculate the end time by adding 30 minutes to `pickup_rounded` using `+ pd.Timedelta(minutes=30)` and format it in the same way.
- The start and end times are concatenated with a `"-"` separator to define the interval (e.g., `02:00-02:30`).

In [14]:
pickup_rounded = df["tpep_pickup_datetime"].dt.floor("30min")

df["pickup_30min_interval"] = (
    pickup_rounded.dt.strftime("%H:%M")
    + "-"
    + (pickup_rounded + pd.Timedelta(minutes=30)).dt.strftime("%H:%M")
)

df[["tpep_pickup_datetime", "pickup_30min_interval"]].head()

tpep_pickup_datetime pickup_30min_interval
0  2020-01-01 00:28:15           00:00-00:30
1  2020-01-01 00:35:39           00:30-01:00
2  2020-01-01 00:47:41           00:30-01:00
3  2020-01-01 00:55:23           00:30-01:00
4  2020-01-01 00:01:58           00:00-00:30

#### 8. For each interval, determine the average number of passengers and the average fare amount.



I group the dataset by `pickup_30min_interval` using `.groupby()` and calculate the following metrics for each interval:
- `avg_passengers`: computed as the mean of the `passenger_count` column.
- `avg_fare_amount`: computed as the mean of the `fare_amount` column.

These metrics are calculated using the `.agg()` method, which allows applying aggregation functions (in this case, `"mean"`) to multiple columns.

In [15]:
interval_stats = df.groupby("pickup_30min_interval").agg(
    avg_passengers=("passenger_count", "mean"),
    avg_fare_amount=("fare_amount", "mean")
).reset_index()

interval_stats.head()

pickup_30min_interval  avg_passengers  avg_fare_amount
0           00:00-00:30        1.572848        13.526433
1           00:30-01:00        1.584345        13.214132
2           01:00-01:30        1.578933        12.699554
3           01:30-02:00        1.589182        12.265997
4           02:00-02:30        1.587479        12.089669

In [16]:
interval_stats.tail()

pickup_30min_interval  avg_passengers  avg_fare_amount
43           21:30-22:00        1.564563        12.559674
44           22:00-22:30        1.565824        12.805710
45           22:30-23:00        1.571608        13.170497
46           23:00-23:30        1.569022        13.253366
47           23:30-00:00        1.570120        13.206596

#### 9. For each payment type and each interval, determine the average fare amount


I group the dataset by `payment_type` and `pickup_30min_interval` using `.groupby()` to calculate the average fare amount (`avg_fare_amount`) for each combination of payment type and interval. 

The aggregation is performed with the `.agg()` method, applying the `"mean"` function to the `fare_amount` column.

In [17]:
payment_interval_stats = df.groupby(["payment_type", "pickup_30min_interval"]).agg(
    avg_fare_amount=("fare_amount", "mean")
).reset_index()

payment_interval_stats.head()


payment_type pickup_30min_interval  avg_fare_amount
0           1.0           00:00-00:30        13.869142
1           1.0           00:30-01:00        13.472232
2           1.0           01:00-01:30        12.824603
3           1.0           01:30-02:00        12.357974
4           1.0           02:00-02:30        12.008589

#### 10. For each payment type, determine the interval when the average fare amount is maximum


I first group the dataset by `payment_type` and `pickup_30min_interval` using `.groupby()` and calculate the average fare amount (`avg_fare_amount`) for each combination. The results are stored in the `payment_interval_stats` DataFrame.

To find the interval with the maximum average fare amount for each `payment_type`, I:
1. Use `.groupby("payment_type")["avg_fare_amount"].idxmax()` to identify the row index of the maximum `avg_fare_amount` for each `payment_type`.
2. Use `.loc[]` to select these rows from the `payment_interval_stats` DataFrame.

In [18]:
payment_interval_stats = df.groupby(["payment_type", "pickup_30min_interval"]).agg(
    avg_fare_amount=("fare_amount", "mean")
).reset_index()

max_fare_intervals = payment_interval_stats.loc[
    payment_interval_stats.groupby("payment_type")["avg_fare_amount"].idxmax()
]

max_fare_intervals


payment_type pickup_30min_interval  avg_fare_amount
10            1.0           05:00-05:30        21.256949
58            2.0           05:00-05:30        14.846814
110           3.0           07:00-07:30        10.950938
154           4.0           05:00-05:30         6.634043
192           5.0           17:30-18:00         0.000000

#### 11. For each payment type, determine the interval when the overall ratio between the tip and the fare amounts is maximum


I first group the dataset by `payment_type` and `pickup_30min_interval` using `.groupby()` and calculate:
- `total_tip`: The sum of `tip_amount` for each group.
- `total_fare`: The sum of `fare_amount` for each group.

The results are stored in the `tip_fare_ratio_stats` DataFrame. 

I then compute the tip-to-fare ratio by dividing `total_tip` by `total_fare` and storing the result in a new column, `tip_to_fare_ratio`. To handle cases where `total_fare` is zero, I fill any resulting `NaN` values with `-1` using `.fillna(-1)`.

To find the interval with the maximum tip-to-fare ratio for each `payment_type`, I:
1. Use `.groupby("payment_type")["tip_to_fare_ratio"].idxmax()` to identify the row index of the maximum `tip_to_fare_ratio` for each `payment_type`.
2. Use `.loc[]` to select these rows from the `tip_fare_ratio_stats` DataFrame.


In [19]:
tip_fare_ratio_stats = df.groupby(["payment_type", "pickup_30min_interval"]).agg(
    total_tip=("tip_amount", "sum"),
    total_fare=("fare_amount", "sum")
).reset_index()

tip_fare_ratio_stats["tip_to_fare_ratio"] = tip_fare_ratio_stats["total_tip"] / tip_fare_ratio_stats["total_fare"]

tip_fare_ratio_stats["tip_to_fare_ratio"] = tip_fare_ratio_stats["tip_to_fare_ratio"].fillna(-1)

max_ratio_intervals = tip_fare_ratio_stats.loc[
    tip_fare_ratio_stats.groupby("payment_type")["tip_to_fare_ratio"].idxmax()
]

max_ratio_intervals


payment_type pickup_30min_interval  total_tip  total_fare  \
37            1.0           18:30-19:00  485543.67  1998355.20   
58            2.0           05:00-05:30      15.00   109406.17   
138           3.0           21:00-21:30      35.62     5646.67   
170           4.0           13:00-13:30      36.48      170.05   
192           5.0           17:30-18:00       0.00        0.00   

     tip_to_fare_ratio  
37            0.242972  
58            0.000137  
138           0.006308  
170           0.214525  
192          -1.000000

In the last row of the result, `total_fare` is `0`, which means the `tip_to_fare_ratio` is set to `-1` as per the handling logic. 






#### 12. Find the location with the highest average fare amount


I first group the dataset by `PULocationID` using `.groupby()` and calculate the average fare amount (`avg_fare_amount`) for each location by applying the `"mean"` aggregation to the `fare_amount` column. The results are stored in the `avg_fare_by_location` DataFrame.

To identify the location with the highest average fare amount:
1. I use `.idxmax()` on the `avg_fare_amount` column to find the row index corresponding to the maximum value.
2. I use `.loc[]` to extract this row from `avg_fare_by_location` and store it in `max_fare_location`.

In [20]:
avg_fare_by_location = df.groupby("PULocationID").agg(
    avg_fare_amount=("fare_amount", "mean")
).reset_index()

max_fare_location = avg_fare_by_location.loc[avg_fare_by_location["avg_fare_amount"].idxmax()]

max_fare_location

PULocationID       204.0
avg_fare_amount    107.0
Name: 198, dtype: float64

#### 13. Build a new dataframe (called common) where, for each pickup location we keep all trips to the 5 most common destinations (i.e. each pickup location can have different common destinations).


I group the dataset by `PULocationID` and `DOLocationID` using `.groupby()` and count the number of trips for each pair with `.size()`. This creates a DataFrame (`top_destinations`) with columns `PULocationID`, `DOLocationID`, and `trip_count`. I then sort the results by `PULocationID` in ascending order and `trip_count` in descending order to identify the most common destinations for each pickup location.

Next, I extract the top 5 destinations for each pickup location by using `.groupby("PULocationID").head(5)`. This step produces a DataFrame (`top_5_destinations`) containing the most frequent destinations for each `PULocationID`.

Finally, I merge the original DataFrame (`df`) with `top_5_destinations` on `PULocationID` and `DOLocationID` using an `inner` join. 


In [21]:
top_destinations = (
    df.groupby(["PULocationID", "DOLocationID"])
    .size()
    .reset_index(name="trip_count")
    .sort_values(["PULocationID", "trip_count"], ascending=[True, False])
)

top_5_destinations = (
    top_destinations.groupby("PULocationID")
    .head(5)
    .reset_index(drop=True)
)

common = df.merge(
    top_5_destinations[["PULocationID", "DOLocationID"]],
    on=["PULocationID", "DOLocationID"],
    how="inner"
)

common.head()


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   
1       1.0  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0   
2       1.0  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0   
3       1.0  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0   
4       2.0  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            1.2         1.0                  N           238           239   
1            1.2         1.0                  N           239           238   
2            0.6         1.0                  N           238           238   
3            0.8         1.0                  N           238           151   
4            0.0         1.0                  N           193           193   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           1.0          6.0    3.0      0.5        1.47           0.0   
1           1.0          7.0    3.0      0.5        1.50           0.0   
2           1.0          6.0    3.0      0.5        1.00           0.0   
3           1.0          5.5    0.5      0.5        1.36           0.0   
4           2.0          3.5    0.5      0.5        0.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  duration  \
0                    0.3         11.27                   2.5  4.800000   
1                    0.3         12.30                   2.5  7.416667   
2                    0.3         10.80                   2.5  6.183333   
3                    0.3          8.16                   0.0  4.850000   
4                    0.3          4.80                   0.0  2.300000   

  pickup_30min_interval  
0           00:00-00:30  
1           00:30-01:00  
2           00:30-01:00  
3           00:30-01:00  
4           00:00-00:30

#### 14. On the common dataframe, for each payment type and each interval, determine the average fare amount


I group the `common` DataFrame by `payment_type` and `pickup_30min_interval` using `.groupby()` to calculate the average fare amount (`avg_fare_amount`) for each combination of payment type and interval. 

I use the `.agg()` function to apply the `"mean"` aggregation to the `fare_amount` column, resulting in the average fare amount for each group.

In [22]:
avg_fare_common = common.groupby(["payment_type", "pickup_30min_interval"]).agg(
    avg_fare_amount=("fare_amount", "mean")
).reset_index()

avg_fare_common.head()


payment_type pickup_30min_interval  avg_fare_amount
0           1.0           00:00-00:30         8.588957
1           1.0           00:30-01:00         8.690681
2           1.0           01:00-01:30         8.496143
3           1.0           01:30-02:00         8.026808
4           1.0           02:00-02:30         7.945909

#### 15. Compute the difference of the average fare amount computed in the previous point with those computed at point 9.


I merge the DataFrame `payment_interval_stats` (computed at point 9) with the DataFrame `avg_fare_common` (computed at point 14) using `.merge()`. The merge is performed on the shared columns `payment_type` and `pickup_30min_interval`, ensuring alignment between the two datasets. I use the `suffixes` parameter to differentiate between the average fare amounts in the merged DataFrame:
- Columns from `payment_interval_stats` are suffixed with `_all`.
- Columns from `avg_fare_common` are suffixed with `_common`.

Next, I calculate the difference between the average fare amounts:
- `avg_fare_amount_common` (from the `common` DataFrame) and 
- `avg_fare_amount_all` (from the original DataFrame).
The result is stored in a new column, `fare_amount_difference`.

In [23]:
merged_fares = payment_interval_stats.merge(
    avg_fare_common,
    on=["payment_type", "pickup_30min_interval"],
    suffixes=("_all", "_common")
)

merged_fares["fare_amount_difference"] = (
    merged_fares["avg_fare_amount_common"] - merged_fares["avg_fare_amount_all"]
)

merged_fares[["payment_type", "pickup_30min_interval", "avg_fare_amount_all", "avg_fare_amount_common", "fare_amount_difference"]].head()


payment_type pickup_30min_interval  avg_fare_amount_all  \
0           1.0           00:00-00:30            13.869142   
1           1.0           00:30-01:00            13.472232   
2           1.0           01:00-01:30            12.824603   
3           1.0           01:30-02:00            12.357974   
4           1.0           02:00-02:30            12.008589   

   avg_fare_amount_common  fare_amount_difference  
0                8.588957               -5.280185  
1                8.690681               -4.781551  
2                8.496143               -4.328460  
3                8.026808               -4.331167  
4                7.945909               -4.062679

#### 16. Compute the ratio between the differences computed in the previous point and those computed in point 9. Note: you have to compute a ratio for each pair (payment type, interval).


I take the DataFrame from the previous step and calculate the ratio by dividing `fare_amount_difference` by `avg_fare_amount_all`. The result is stored in the `fare_difference_ratio` column.


In [24]:
merged_fares["fare_difference_ratio"] = (
    merged_fares["fare_amount_difference"] / merged_fares["avg_fare_amount_all"]
)

merged_fares[[
    "payment_type",
    "pickup_30min_interval",
    "fare_amount_difference",
    "avg_fare_amount_all",
    "fare_difference_ratio"
]].head()


payment_type pickup_30min_interval  fare_amount_difference  \
0           1.0           00:00-00:30               -5.280185   
1           1.0           00:30-01:00               -4.781551   
2           1.0           01:00-01:30               -4.328460   
3           1.0           01:30-02:00               -4.331167   
4           1.0           02:00-02:30               -4.062679   

   avg_fare_amount_all  fare_difference_ratio  
0            13.869142              -0.380715  
1            13.472232              -0.354919  
2            12.824603              -0.337512  
3            12.357974              -0.350475  
4            12.008589              -0.338314

#### 17. Build chains of trips. Two trips are consecutive in a chain if (a) they have the same VendorID, (b) the pickup location of the second trip is also the dropoff location of the first trip, (c) the pickup time of the second trip is after the dropoff time of the first trip, and (d) the pickup time of the second trip is at most 2 minutes later than the dropoff time of the first trip. Hint: Add a column chain to the dataset. A chain can have more than two trips.




To process the dataset, I start by converting the date columns into a proper datetime format, as this allows for easier manipulation later on. The data is then sorted by `VendorID` and `tpep_pickup_datetime` to ensure that the trips are processed in a logical order.

The main objective here is to assign chain IDs to trips, grouping those that are linked by matching drop-off and pick-up locations within a time window of 120 seconds. To achieve this efficiently, I convert the relevant columns into NumPy arrays, which allow for faster operations compared to direct DataFrame manipulations.

For each trip, I determine whether it already belongs to a chain. If not, I assign a new chain ID. Using binary search (`bisect_left` and `bisect_right`), I narrow down the range of potential matching trips based on the 120-second window. Within this range, I then iterate over the trips to check if the `PULocationID` matches the `DOLocationID` of the current trip. If a match is found, I propagate the chain ID to the matching trip and stop further checks using a `break`. This ensures efficiency and also resolves cases where multiple potential matches exist by assuming the closest trip in time is the correct chain match.

Finally, I update the `chain` column in the original DataFrame with the computed chain IDs and sort the data by chain for a clear view of the grouped trips.


In [25]:
import pandas as pd
from tqdm import tqdm
import numpy as np
from bisect import bisect_left, bisect_right



# Convert datetime columns
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df['tpep_dropoff_datetime'] = pd.to_datetime(df['tpep_dropoff_datetime'])

# Sort the DataFrame
df = df.sort_values(by=["VendorID", "tpep_pickup_datetime"]).reset_index(drop=True)

# Initialize the `chain` column to 0
df['chain'] = 0

# Variable to track the current chain_id
current_chain_id = 1

# Convert relevant columns to NumPy arrays
vendor_ids = df['VendorID'].values
pickup_times = df['tpep_pickup_datetime'].astype(np.int64) // 10**9  # Convert to UNIX seconds
dropoff_times = df['tpep_dropoff_datetime'].astype(np.int64) // 10**9  # Convert to UNIX seconds
pu_locations = df['PULocationID'].values
do_locations = df['DOLocationID'].values
chain_array = np.zeros(len(df), dtype=np.int32)  # NumPy array to track chain_ids

# Iterate over the rows with a time window
for i in tqdm(range(len(df)), desc="Assigning chains"):
    # Assign a new chain if the current trip does not have one
    if chain_array[i] == 0:
        chain_array[i] = current_chain_id
        current_chain_id += 1

    # Get the current VendorID and dropoff time
    current_vendor = vendor_ids[i]
    current_dropoff_time = dropoff_times[i]
    current_chain_id_local = chain_array[i]

    # Find the range of relevant rows using binary search
    start_index = bisect_left(pickup_times, current_dropoff_time, lo=i+1)
    end_index = bisect_right(pickup_times, current_dropoff_time + 120, lo=i+1)

    # Iterate only over the relevant rows
    for j in range(start_index, end_index):
        # Skip if the VendorID is different
        if vendor_ids[j] != current_vendor:
            break

        # Check if the conditions for matching are met
        if (
            pu_locations[j] == do_locations[i]
        ):
            # Assign the same chain_id
            chain_array[j] = current_chain_id_local
            break  # Exit the loop when the first match is found

# Update the DataFrame with the chain values
df['chain'] = chain_array

# Sort by chain and display results
df_sorted_by_chain = df.sort_values(by=["chain", "tpep_pickup_datetime"])
print(df_sorted_by_chain[[
    "VendorID", "tpep_pickup_datetime", "tpep_dropoff_datetime", "PULocationID", "DOLocationID", "chain"
]].head(20))


Assigning chains: 100%|██████████| 6339567/6339567 [34:52<00:00, 3030.22it/s] 


      VendorID tpep_pickup_datetime tpep_dropoff_datetime  PULocationID  \
0          1.0  2020-01-01 00:00:00   2020-01-01 00:13:03            68   
538        1.0  2020-01-01 00:13:28   2020-01-01 00:31:48           170   
1871       1.0  2020-01-01 00:31:54   2020-01-01 01:42:07           239   
1          1.0  2020-01-01 00:00:03   2020-01-01 00:13:04            79   
509        1.0  2020-01-01 00:13:04   2020-01-01 00:17:41           162   
868        1.0  2020-01-01 00:18:23   2020-01-01 00:30:29           162   
1790       1.0  2020-01-01 00:30:51   2020-01-01 00:52:56           161   
2          1.0  2020-01-01 00:00:05   2020-01-01 00:26:30           264   
3          1.0  2020-01-01 00:00:07   2020-01-01 00:03:26            75   
4          1.0  2020-01-01 00:00:25   2020-01-01 00:05:59           145   
5          1.0  2020-01-01 00:00:26   2020-01-01 00:08:31           113   
6          1.0  2020-01-01 00:00:41   2020-01-01 00:16:23            90   
7          1.0  2020-01-0

In [26]:
chain_counts = df["chain"].value_counts().reset_index()
chain_counts.columns = ["chain", "trip_count"]

multi_trip_chains = chain_counts[chain_counts["trip_count"] > 5]

chains_with_details = df.merge(multi_trip_chains, on="chain")

chains_with_details_sorted = chains_with_details.sort_values(by=["chain", "tpep_pickup_datetime"])
chains_with_details_sorted[["VendorID", "tpep_pickup_datetime", "tpep_dropoff_datetime", "PULocationID", "DOLocationID", "chain"]].head(20)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  PULocationID  \
0         1.0  2020-01-01 00:03:02   2020-01-01 00:09:22           232   
5         1.0  2020-01-01 00:09:36   2020-01-01 00:12:57            79   
6         1.0  2020-01-01 00:13:00   2020-01-01 00:19:15            79   
15        1.0  2020-01-01 00:19:42   2020-01-01 00:27:10           234   
31        1.0  2020-01-01 00:27:42   2020-01-01 00:36:23           249   
55        1.0  2020-01-01 00:36:24   2020-01-01 00:47:12            79   
1         1.0  2020-01-01 00:08:15   2020-01-01 00:18:12           170   
13        1.0  2020-01-01 00:18:44   2020-01-01 00:22:36           263   
21        1.0  2020-01-01 00:23:32   2020-01-01 00:28:46           262   
34        1.0  2020-01-01 00:28:46   2020-01-01 00:49:34           229   
101       1.0  2020-01-01 00:49:38   2020-01-01 00:58:21            68   
138       1.0  2020-01-01 00:58:42   2020-01-01 01:38:52            68   
321       1.0  2020-01-01 01:39:18   2020-01-01 02:04:20           181   
2         1.0  2020-01-01 00:08:24   2020-01-01 00:29:31           151   
36        1.0  2020-01-01 00:29:37   2020-01-01 00:35:37           234   
54        1.0  2020-01-01 00:35:50   2020-01-01 00:42:04           164   
75        1.0  2020-01-01 00:42:10   2020-01-01 00:52:35           170   
113       1.0  2020-01-01 00:52:37   2020-01-01 01:10:09           229   
188       1.0  2020-01-01 01:10:23   2020-01-01 01:17:28           234   
3         1.0  2020-01-01 00:08:29   2020-01-01 00:13:20           114   

     DOLocationID  chain  
0              79     56  
5              79     56  
6             234     56  
15            249     56  
31             79     56  
55             87     56  
1             263    215  
13            262    215  
21            229    215  
34             68    215  
101            68    215  
138           181    215  
321            37    215  
2             234    217  
36            164    217  
54            170    217  
75            229    217  
113           234    217  
188           249    217  
3             148    220